src/audio/copier/copier_gain.c

```diff
@@ -0,0 +1,192 @@
+		return -EINVAL;
+	}
+}
+
+enum copier_gain_state copier_gain_eval_state(struct copier_gain_params *gain_params)
+{
+	enum copier_gain_state state = STATIC_GAIN;
+
+	if (gain_params->silence_sg_count < gain_params->silence_sg_length)
+		state = TRANS_MUTE;
+	else if ((gain_params->fade_in_sg_count < gain_params->fade_sg_length) &&
+		 (gain_params->fade_sg_length != 0))
+		state = TRANS_GAIN;
+
+	return state;
+}
+
+int copier_gain_dma_control(uint32_t node_id, const uint32_t *config_data,
+			    size_t config_size, enum sof_ipc_dai_type dai_type)
+{
+	union ipc4_connector_node_id node = (union ipc4_connector_node_id)node_id;
+	struct sof_tlv *tlv = (struct sof_tlv *)config_data;
+	struct gain_dma_control_data *gain_data = NULL;
+	struct ipc_comp_dev *icd = NULL;
+	struct processing_module *mod = NULL;
+	struct copier_data *cd = NULL;
+	struct ipc *ipc = ipc_get();
+	struct list_item *clist;
+	struct comp_dev *dev;
+	int ret;
+
+	list_for_item(clist, &ipc->comp_list) {
+		icd = container_of(clist, struct ipc_comp_dev, list);
+
+		if (icd->type != COMP_TYPE_COMPONENT)
+			continue;
+
+		dev = icd->cd;
+
+		if (dev->ipc_config.type != SOF_COMP_DAI)
+			continue;
+
+		mod = comp_mod(dev);
+		cd = module_get_private_data(mod);
+
+		switch (dai_type) {
+		case SOF_DAI_INTEL_DMIC:
+			if (cd->dd[0]->dai->index != node.f.v_index)
+				continue;
+
+			if (!config_size) {
+				comp_err(dev, "Config length for DMIC couldn't be zero");
+				return -EINVAL;
+			}
+
+			/* Gain coefficients for DMIC */
+			void *tlv_val = tlv_value_ptr_get(tlv, DMIC_SET_GAIN_COEFFICIENTS);
+			if (!tlv_val) {
+				comp_err(dev, "No gain coefficients in DMA_CONTROL ipc");
+				return -EINVAL;
+			}
+			gain_data = tlv_val;
+			break;
+		default:
+			comp_warn(dev, "Gain DMA control: no dai type=%d found", dai_type);
+			break;
+		}
+
+		ret = copier_set_gain(dev, cd->dd[0], gain_data);
+		if (ret)
+			comp_err(dev, "Gain DMA control: failed to set gain");
+
+		return ret;
+	}
+
+	comp_err(dev, "Gain DMA control: no dai component found dai type=%d", dai_type);
+	return -ENODEV;
+}
+
+int copier_set_gain(struct comp_dev *dev, struct dai_data *dd,
+		    struct gain_dma_control_data *gain_data)
+{
+	struct copier_gain_params *gain_params = dd->gain_data;
+	struct ipc4_copier_module_cfg *copier_cfg = dd->dai_spec_config;
+	uint16_t static_gain[MAX_GAIN_COEFFS_CNT];
+	int ret;
+
+	if (!gain_data) {
+		comp_err(dev, "Gain data is NULL");
+		return -EINVAL;
+	}
+
+	/* Set gain coefficients */
+	comp_info(dev, "Update gain coefficients from DMA_CONTROL ipc");
+
+	ret = memcpy_s(static_gain, sizeof(static_gain), gain_data->gain_coeffs,
+		       sizeof(gain_data->gain_coeffs));
+	assert(!ret);
+
+	for (size_t i = 0; i < MAX_GAIN_COEFFS_CNT; i++)
+		static_gain[i] = static_gain[i % copier_cfg->base.audio_fmt.channels_count];
+
+	ret = memcpy_s(gain_params->gain_coeffs, sizeof(gain_params->gain_coeffs),
+		       static_gain, sizeof(static_gain));
+	assert(!ret);
```

Comment:
An `assert()` is something that can be and is usually disabled in production[*]. So `asserts` are only for checking things that can never happen in theory == bugs. Please add a short comment explaining why this `memcpy_s()` can never fail.


From https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1969.htm
> Testing Return Values For Errors
Finally, as mentioned above, the need to test for and handle errors reported by functions that historically could not fail is typically viewed as either unnecessary, or overly intrusive and difficult, and tends to be omitted without an understanding of the consequences of allowing the affected code to continue to run after a runtime-constraint violation.

[*] quoting for instance https://github.com/thesofproject/sof/issues/9308#issuecomment-2244736317
> CI caught the problem only in builds with assertions enabled, but also in a normal build, the heap initialization was also overwriting our FW, it just didn't cause crashes.

Reply:
Alternatively:
```C
 if (memcpy_s(gain_params->gain_coeffs, sizeof(gain_params->gain_coeffs),
		       static_gain, sizeof(static_gain)))
        return -ESOMETHING;
```